### ISSUE TITLE
TypeError in `check_executable_overwrite`

### ISSUE THREAD

**Issue by joallard** [rbenv/ruby-build#1014]
Invocation:
```
$ RUBY_CONFIGURE_OPTS=--with-readline-dir="$(brew --prefix readline)" rbenv install 2.3.1
rbenv: /Users/jon/.rbenv/versions/2.3.1 already exists
continue with installation? (y/N) y
Downloading ruby-2.3.1.tar.bz2...
-> https://cache.ruby-lang.org/pub/ruby/2.3/ruby-2.3.1.tar.bz2
Installing ruby-2.3.1...

BUILD FAILED (OS X 10.11.6 using ruby-build 20160913-13-g8ef0c34)

Inspect or clean up the working tree at /var/folders/p4/n654g1ds69z47bvnzwc7bq000000gn/T/ruby-build.20161103153651.91718
Results logged to /var/folders/p4/n654g1ds69z47bvnzwc7bq000000gn/T/ruby-build.20161103153651.91718.log

Last 10 log lines:
	from /private/var/folders/p4/n654g1ds69z47bvnzwc7bq000000gn/T/ruby-build.20161103153651.91718/ruby-2.3.1/lib/rubygems/specification.rb:742:in `each'
	from /private/var/folders/p4/n654g1ds69z47bvnzwc7bq000000gn/T/ruby-build.20161103153651.91718/ruby-2.3.1/lib/rubygems/specification.rb:742:in `block in each_gemspec'
	from /private/var/folders/p4/n654g1ds69z47bvnzwc7bq000000gn/T/ruby-build.20161103153651.91718/ruby-2.3.1/lib/rubygems/specification.rb:741:in `each'
	from /private/var/folders/p4/n654g1ds69z47bvnzwc7bq000000gn/T/ruby-build.20161103153651.91718/ruby-2.3.1/lib/rubygems/specification.rb:741:in `each_gemspec'
	from /private/var/folders/p4/n654g1ds69z47bvnzwc7bq000000gn/T/ruby-build.20161103153651.91718/ruby-2.3.1/lib/rubygems/specification.rb:819:in `each_spec'
	from ./tool/rbinstall.rb:751:in `block in <main>'
	from ./tool/rbinstall.rb:801:in `block in <main>'
	from ./tool/rbinstall.rb:798:in `each'
	from ./tool/rbinstall.rb:798:in `<main>'
make: *** [do-install-all] Error 1
```

Logfile tail:

```
generating x86_64-darwin15-fake.rb
x86_64-darwin15-fake.rb updated
./miniruby -I./lib -I. -I.ext/common  ./tool/runruby.rb --extout=.ext  -- --disable-gems -r./x86_64-darwin15-fake ./tool/rbinstall.rb --make="make" --dest-dir="" --extout=".ext" --mflags="" --make-flags="" --data-mode=0644 --prog-mode=0755 --installed-list .installed.list --mantype="doc" --install=all --rdoc-output=".ext/rdoc"
installing binary commands:   /Users/jon/.rbenv/versions/2.3.1/bin
installing base libraries:    /Users/jon/.rbenv/versions/2.3.1/lib
installing arch files:        /Users/jon/.rbenv/versions/2.3.1/lib/ruby/2.3.0/x86_64-darwin15
installing pkgconfig data:    /Users/jon/.rbenv/versions/2.3.1/lib/pkgconfig
installing command scripts:   /Users/jon/.rbenv/versions/2.3.1/bin
installing library scripts:   /Users/jon/.rbenv/versions/2.3.1/lib/ruby/2.3.0
installing common headers:    /Users/jon/.rbenv/versions/2.3.1/include/ruby-2.3.0
installing manpages:          /Users/jon/.rbenv/versions/2.3.1/share/man/man1
installing extension objects: /Users/jon/.rbenv/versions/2.3.1/lib/ruby/2.3.0/x86_64-darwin15
installing extension objects: /Users/jon/.rbenv/versions/2.3.1/lib/ruby/site_ruby/2.3.0/x86_64-darwin15
installing extension objects: /Users/jon/.rbenv/versions/2.3.1/lib/ruby/vendor_ruby/2.3.0/x86_64-darwin15
installing extension headers: /Users/jon/.rbenv/versions/2.3.1/include/ruby-2.3.0/x86_64-darwin15
installing extension scripts: /Users/jon/.rbenv/versions/2.3.1/lib/ruby/2.3.0
installing extension scripts: /Users/jon/.rbenv/versions/2.3.1/lib/ruby/site_ruby/2.3.0
installing extension scripts: /Users/jon/.rbenv/versions/2.3.1/lib/ruby/vendor_ruby/2.3.0
installing extension headers: /Users/jon/.rbenv/versions/2.3.1/include/ruby-2.3.0/ruby
installing default gems:      /Users/jon/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0 (build_info, cache, doc, extensions, gems, specifications)
                              bigdecimal 1.2.8
                              io-console 0.4.5
                              json 1.8.3
                              psych 2.0.17
                              rdoc 4.2.1
installing bundle gems:       /Users/jon/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0 (build_info, cache, doc, extensions, gems, specifications)
                              did_you_mean 1.0.0
                              minitest 5.8.3
                              net-telnet 0.1.1
                              power_assert 0.2.6
                              rake 10.4.2
/private/var/folders/p4/n654g1ds69z47bvnzwc7bq000000gn/T/ruby-build.20161103153651.91718/ruby-2.3.1/lib/rubygems/installer.rb:227:in `check_executable_overwrite': no implicit conversion of nil into String (TypeError)
	from /private/var/folders/p4/n654g1ds69z47bvnzwc7bq000000gn/T/ruby-build.20161103153651.91718/ruby-2.3.1/lib/rubygems/installer.rb:468:in `block in generate_bin'
	from /private/var/folders/p4/n654g1ds69z47bvnzwc7bq000000gn/T/ruby-build.20161103153651.91718/ruby-2.3.1/lib/rubygems/installer.rb:455:in `each'
	from /private/var/folders/p4/n654g1ds69z47bvnzwc7bq000000gn/T/ruby-build.20161103153651.91718/ruby-2.3.1/lib/rubygems/installer.rb:455:in `generate_bin'
	from /private/var/folders/p4/n654g1ds69z47bvnzwc7bq000000gn/T/ruby-build.20161103153651.91718/ruby-2.3.1/lib/rubygems/installer.rb:295:in `install'
	from ./tool/rbinstall.rb:686:in `call'
	from ./tool/rbinstall.rb:686:in `block in <class:Installer>'
	from ./tool/rbinstall.rb:754:in `block (2 levels) in <main>'
	from /private/var/folders/p4/n654g1ds69z47bvnzwc7bq000000gn/T/ruby-build.20161103153651.91718/ruby-2.3.1/lib/rubygems/specification.rb:821:in `block in each_spec'
	from /private/var/folders/p4/n654g1ds69z47bvnzwc7bq000000gn/T/ruby-build.20161103153651.91718/ruby-2.3.1/lib/rubygems/specification.rb:743:in `block (2 levels) in each_gemspec'
	from /private/var/folders/p4/n654g1ds69z47bvnzwc7bq000000gn/T/ruby-build.20161103153651.91718/ruby-2.3.1/lib/rubygems/specification.rb:742:in `each'
	from /private/var/folders/p4/n654g1ds69z47bvnzwc7bq000000gn/T/ruby-build.20161103153651.91718/ruby-2.3.1/lib/rubygems/specification.rb:742:in `block in each_gemspec'
	from /private/var/folders/p4/n654g1ds69z47bvnzwc7bq000000gn/T/ruby-build.20161103153651.91718/ruby-2.3.1/lib/rubygems/specification.rb:741:in `each'
	from /private/var/folders/p4/n654g1ds69z47bvnzwc7bq000000gn/T/ruby-build.20161103153651.91718/ruby-2.3.1/lib/rubygems/specification.rb:741:in `each_gemspec'
	from /private/var/folders/p4/n654g1ds69z47bvnzwc7bq000000gn/T/ruby-build.20161103153651.91718/ruby-2.3.1/lib/rubygems/specification.rb:819:in `each_spec'
	from ./tool/rbinstall.rb:751:in `block in <main>'
	from ./tool/rbinstall.rb:801:in `block in <main>'
	from ./tool/rbinstall.rb:798:in `each'
	from ./tool/rbinstall.rb:798:in `<main>'
make: *** [do-install-all] Error 1
```

What's weird is that I previously installed 2.3.1 with ruby-build just fine, I just had some problems with Readline and wanted to rebuild.

**Comment by joallard** [comment]
Seems to be related to rubygems/rubygems#1557 and rubygems/rubygems#1561, though the PR was merged in 2.6.3 and I had Gem 2.6.5 before upgrading.

**Comment by mislav** [comment]
If you clear the installation directory and try again, does it go away?

**Comment by joallard** [comment]
Not sure if that changes anything, but in between trying different things (and freaking out), I've upgraded my current rubygems. The install worked after doing that and installing on a clean dir.

**Comment by mislav** [comment]
It's weird that an installation of Ruby would be affected by your current rubygems.

**Comment by joallard** [comment]
It is very weird indeed, but I'm not sure which one of the two was the culprit. I'll close since I don't need any more help for this, hopefully if someone comes up with the same issue it can help them. It's probably a very edgy case.

**Comment by mislav** [comment]
Thanks for reporting!

**Comment by Mazwak** [comment]
I've ran into the issue also.

It's a rubygem bug. Look at :

https://github.com/rubygems/rubygems/commit/a8e9b83c227f54da077d9623b6a86c5589dd40f6
https://github.com/rubygems/rubygems/commit/13afe08ec7dfcbbc09e7d76aedb476c0694c05fd
https://github.com/rubygems/rubygems/commit/089a89316ec416632e6a07ffe3567fda6ee6e0f2

The bug triggers only for rake afaict.

**Comment by joallard** [comment]
@Mazwak Interesting, I kinda expected a "new" install to provide its own new-ish version of Rubygems and use that to do the install.

**Comment by Mazwak** [comment]
That's exactly the problem. When I install 2.3.1 over itself, it uses the bugged version that was released with 2.3.1, instead of the fixed latest version.
In your terms, the "new" installation provides the old-ish version.
